File: code/functions/rprism/R/util_load.R
#' Delineate treatments from compact format to more user-friendly format
#'
#' Compressed data must consist of a name with additional data enclosed in brackets and separated by a common delimiter. 
#'
#' @param df a data.frame that must contain the column \code{col_name}.
#' @param col_name Name of the column to be delineated.
#' @param format Character vector of delineated column names corresponding the format of the compressed data.
#' @param separate_rows (optional) Should data be first unnested on the "|" separator"?
#' @param prefix (optional) Prefix to the new columns that contain the delineated data.
#' @param suffix (optional) Suffix to the new columns that contain the delineated data.
#' @param delineate_dates (optional) Used only if "Dates" is in \code{format}. Split "Dates" into "Date_Min" and
#'   "Date_Max".
#' @param convert_dates_to_datetime (optional) Used only if delineate_dates=True. Should dates be converted to datetime
#'   type?
#' @param sep (optional) The delimiter of the compressed data.
#' @param name (optional) Name given to the column containing the value before the brackets. Prefix and suffixes
#' specified will be added to this name.
#' @return a data.frame with possibly more rows (if \code{separate_rows} is set to \code{TRUE}) and more columns
#'   (as many as names in the list \code{format}) built from \code{df}.
#'
#' @importFrom rlang .data
#' @importFrom magrittr %>%
#' @importFrom tidyr separate_rows separate
#'
#' @export
delineate_data <- function(df, col_name, format, separate_rows=T, prefix="", suffix="", delineate_dates=F,
                           convert_dates_to_datetime=T, sep=";", name="Name"){ 

  if (separate_rows){
    df <- df %>% 
      tidyr::separate_rows({{ col_name }}, sep="\\|")
  }

  col_delineate <- "data"
  regex_data <- "(?<=\\()(.*)(?=\\))"

  extract_name <- function(s){
    regex_name <- "^(.*)(?=\\()"
    regex_out <- regexpr(regex_name, s, perl=T)
    if (is.na(regex_out)){
      return(NA)
    } else {
      return(regmatches(s, regex_out))
    }
  }

  df[[name]] <- sapply(df[[col_name]], extract_name, simplify=T, USE.NAMES=F)
  df[[col_delineate]] <- sapply(df[[col_name]], function(s) regmatches(s, regexpr(regex_data, s, perl=T)), USE.NAMES=F)

  format_names <- paste0(prefix, format, suffix)
  suppressWarnings({df <- df %>%
    tidyr::separate(.data[[col_delineate]], format_names, sep=sep)})
  df[[format_names]] <- df[[format_names]] %>% replace(.=="character(0)", NA)

  if (delineate_dates & "Dates" %in% format){
    col_dates <- paste0(prefix, "Dates", suffix)
    col_date_min <- paste0(prefix, "Date_Min", suffix)
    col_date_max <- paste0(prefix, "Date_Min", suffix)
    df <- df %>%
      tidyr::separate({{ col_dates }}, c(col_date_min, col_date_max), sep="_to_")
  }

  if (convert_dates_to_datetime){
    cols_date <- format_names[grepl("Date", format_names) & !grepl("Dates", format_names)]
    df <- df %>%
      dplyr::mutate_at(cols_date, as.Date)
  }

  df
}


#' Load header of a table.
#'
#' @param path Path to the file.
#' @param prefix Prefix for identifying header
#'
#' @keywords internal
read_header <- function(path, prefix){
  hea <- list()
  if (grepl(".gz$", path)){
    con <- gzfile(path, "rt")
  } else {
    con <- file(path, "r")
  }
  i <- 1
  while ( TRUE ) {
    line <- readLines(con, n = 1)
    if ( length(line) == 0 | !grepl(paste0("^", prefix), line)) {
      break
    } else {
      hea[[i]] <- line
      i <- i+1
    }
  }
  close(con)

  hea
}


#' Load a table.
#'
#' @param path Path to the file.
#' @param header_prefix If not NULL, skip header rows that will be identified as first lines starting with this prefix.
#' @param ... Extra parameters passed to \code{read_delim}.
#' 
#' @importFrom readr read_delim
#' @importFrom readxl read_excel
#'
#' @export
load_table <- function(path, header_prefix=NULL, ...){
  if (is.null(path)) return(NULL)
  args <- list(...)

  if (!"delim" %in% names(args)){
    if (grepl(".csv", path)){
      args$delim <- ","
    } else if (grepl(".tsv", path) | grepl(".txt", path)){
      args$delim <- "\t"
    }
  }

  if (!is.null(header_prefix)){
    header <- read_header(path, prefix=header_prefix)
    args$skip <- length(header) 
  }

  if (grepl(".xlsx$", path)){
    df <- do.call(readxl::read_excel, c(path, args))
  } else {
    if (!"progress" %in% names(args)){
      args$progress <- F
    }
    if (!"show_col_types" %in% names(args)){
      args$show_col_types <- F
    }

    if(grepl(".gz$", path)) {
      file <- base::gzfile(path)
      df <- do.call(readr::read_delim, c(list(file=file), args))
    } else {
      df <- do.call(readr::read_delim, c(path, args))
    }
  }

  df
}

#' Load a table located in the data folder
#'
#' @param path Path to the file.
#' @param ... Extra parameters passed to \code{read_delim}.
#' 
#' @importFrom readr read_delim
#' @importFrom readxl read_excel
#'
#' @export
load_from_data <- function(path, ...){
  cwd <- setwd_to_data()
  df <- load_table(path, ...)
  setwd(cwd)

  df
}


#' Write a table to a path in the data folder
#'
#' @param df a data.frame that contains the column \code{col_name}.
#' @param path a path relative to the folder \code{data/}.
#' @param col_name Name of the the column to be used to susbet the dataframe.
#' @param ... Extra parameters passed to \code{write_xlsx} or \code{write_delim}.
#'
#' @importFrom readr write_delim 
#' @importFrom writexl write_xlsx
#'
#' @export
save_to_data <- function(df, path, ...){
  cwd <- setwd_to_data()
  if (grepl(".xlsx$", path)){
    writexl::write_xlsx(df, path, ...)
  } else {
    if (grepl(".tsv$", path)){
      readr::write_delim(df, path, delim="\t", ...)
    } else if (grepl(".csv$", path)){
      readr::write_delim(df, path, delim=",", ...)
    } else {
      readr::write_delim(df, path, ...)
    }
  }
  cat(paste("-file saved at", path, "\n"))
  setwd(cwd)
}


#' Load a table located in the data folder
#'
#' @param df a data.frame that contains the column \code{col_name}.
#' @param values a list of values to be selected.
#' @param col_name Name of the the column to be used to susbet the dataframe.
#' @return a data.frame containing a subset of the rows of \code{df}.
#'
#' @export
subset_data <- function(df, values=NULL, col_name=NULL){
  if (is.null(col_name) | is.null(values)){
    return(df)
  } else {
    return(df[df[[col_name]] %in% values, ,drop=F])
  }
}

File: code/functions/pyprism/pyprism/data/_util_load.py
# -*- coding: utf-8 -*-
"""
@created: Dec 15 2020
@modified: Apr 27 2021
@author: Yoann Pradat

    CentraleSupelec
    MICS laboratory
    9 rue Juliot Curie, Gif-Sur-Yvette, 91190 France

    Institut Gustave Roussy
    Prism Center
    114 rue Edouard Vaillant, Villejuif, 94800 France

Helper functions for loading.

Functions
    def check_filepath_exists_in_data(filepath) -> bool:
    def delineate_data(df, col_name, prefix="", suffix="", format=["Dates", "Cycle", "Dose"], delineate_dates=False
                       convert_dates_to_datetime=True, sep=","):
    def load_table(path: str, sep=None, header=0, dtype=None, **kwargs) -> DataFrame:
    def load_from_data(path: str, sep=None, header=0, dtype=None, **kwargs) -> DataFrame:
    def read_header(path: str) -> list:
    def subset_data(df: DataFrame, values: Iterable[Any]=None, col_name: str=None):
"""

import os
import gzip
import numpy as np
import pandas as pd
from pandas.api.types import is_datetime64_any_dtype
import re
from pyprism.util import setwd_to_data, explode_df
from typing import Iterable, Any, Callable

DataFrame = pd.core.frame.DataFrame


def convert_dates_to_str(df, strftime="%d/%m/%Y"):
    df_c = df.copy()
    for x in df.columns:
        if is_datetime64_any_dtype(df_c[x]):
            df_c.loc[:,x] = df_c[x].apply(lambda x: x.strftime(strftime) if not pd.isnull(x) else "").astype(str)
    return df_c


def check_filepath_exists_in_data(filepath) -> bool:
    if filepath is None:
        filepath_exists = False
    else:
        cwd = setwd_to_data()
        filepath_exists = os.path.exists(filepath)
        os.chdir(cwd)

    return filepath_exists


def delineate_data(df, col_name, prefix="", suffix="", format=["Dates", "Cycle", "Dose"], delineate_dates=False,
                   convert_dates_to_datetime=True, sep=",", name="Name", remove=False, preserve_index=False, **kwargs):
    """
    Delineate treatments from compact format to more user-friendly format.

    Parameters
    ----------
    df: DataFrame
        Dataframe that must contain the column `col_name`.
    col_name: str
        Name of the column to be delineated
    prefix: str, default=""
        Prefix to the new columns that contain the delineated data.
    suffix: str, default=""
        Suffix to the new columns that contain the delineated data.
    format: str, default=["Dates", "Cycle", "Dose"]
        List of delineated column names corresponding the format of the compressed data.
    delineate_dates: str, default=True
        Set to True only if "Dates" is in `format`. Split Dates into Date_Min and Date_Max.
    convert_dates_to_datetime: str, default=True
        Used only if delineate_dates=True. Should dates be converted to datetime type?
    sep: str, default=","
        The delimiter
    name: str, default="Name"
        Name given to the column containing the value before the brackets. Prefix and suffixes specified will be added
        to this name.
    remove: bool, default=False,
        If True, remove the column `col_name`.
    preserve_index: bool, default=False,
        If True, preserve the original index names.
    **kwargs:
        Additional arguments passed to pd.to_datetime function. Used only if convert_dates_to_datetime is True.
    """
    df_c = df.copy()
    df_c[col_name] = df_c[col_name].fillna("Unknown")
    df_c = explode_df(df_c, cols=[col_name], sep="|", preserve_index=preserve_index)

    s_extract = df_c[col_name].str.extract(r"(?<=\()(.*)(?=\))")
    df_c_split = s_extract[0].str.split(sep, expand=True)
    df_c_split.columns = ["%s%s%s" % (prefix, f, suffix) for f in format][:df_c_split.shape[1]]

    if delineate_dates and "Dates" in format:
        col_dates = "%sDates%s" % (prefix,suffix)
        col_date_min = "%sDate_Min%s" % (prefix,suffix)
        col_date_max = "%sDate_Max%s" % (prefix,suffix)
        df_c_split_dates = df_c_split[col_dates].str.split("_to_", expand=True)

        # no value with "_to_" tag
        if df_c_split_dates.shape[1] == 1:
            df_c_split[[col_date_min, col_date_max]] = np.nan
        else:
            df_c_split[[col_date_min, col_date_max]] = df_c_split[col_dates].str.split("_to_", expand=True)

    if convert_dates_to_datetime:
        cols_date = [x for x in df_c_split.columns if "Date" in x and not "Dates" in x]
        for col_date in cols_date:
            df_c_split[col_date] = pd.to_datetime(df_c_split[col_date], **kwargs)

    c_name = "%s%s%s" % (prefix,name,suffix)
    df_c[c_name] = df_c[col_name].apply(lambda x: re.sub(r"\([^)]*\)", "", x))

    df_c[col_name] = df_c[col_name].replace("Unknown", np.nan)
    df_c[c_name] = df_c[c_name].replace("Unknown", np.nan)

    df_c = pd.concat((df_c, df_c_split), axis=1)

    if remove:
        del df_c[col_name]

    return df_c


def _read_excel_with_dates(path, skiprows=None, header=0, dtype=None, cols_date=None, dayfirst=True, format=None, **kwargs):
    if dtype is None and cols_date is not None:
        dtype = {k: "str" for k in cols_date}
    elif dtype is not None and cols_date is not None:
        for col in cols_date:
            dtype[col] = "str"
    elif cols_date is not None:
        raise ValueError("If using both cols_date and dtype arguments, please specify a dictionary for dtype.")

    df_cln = pd.read_excel(path, na_values=["na", "na ", "NA", ".", "not available"], header=header,
                           skiprows=skiprows, parse_dates=False, engine="openpyxl", dtype=dtype, **kwargs)

    if cols_date is not None:
        for col in cols_date:
            if format is not None:
                df_cln.loc[:, col] = pd.to_datetime(df_cln[col], format=format, errors="coerce")
            else:
                df_cln.loc[:, col] = pd.to_datetime(df_cln[col], dayfirst=dayfirst, errors="coerce")

    return df_cln



def load_table(path: str, header=0, dtype=None, header_prefix=None, **kwargs) -> DataFrame:
    if header_prefix is not None:
        skiprows = len(read_header(path=path, prefix=header_prefix))
    else:
        skiprows = None

    if path.endswith("xlsx"):
        df = _read_excel_with_dates(path, skiprows=skiprows, header=header, dtype=dtype, **kwargs)
    elif ".txt" in path or ".tsv" in path or ".maf" in path:
        if "sep" not in kwargs:
            df = pd.read_csv(path, skiprows=skiprows, header=header, dtype=dtype, sep="\t", **kwargs)
        else:
            df = pd.read_csv(path, skiprows=skiprows, header=header, dtype=dtype, **kwargs)
    else:
        df = pd.read_csv(path, skiprows=skiprows, header=header, dtype=dtype, **kwargs)

    return df


def load_from_data(path: str, header=0, dtype=None, header_prefix=None, **kwargs) -> DataFrame:
    cwd = setwd_to_data()
    df = load_table(path, header, dtype, header_prefix, **kwargs)
    os.chdir(cwd)

    return df


def make_column_groups(df: DataFrame, cols2groups) -> DataFrame:
    df_new = df.loc[:, cols2groups.keys()].copy()
    tuples =  [(v,k) for k, v in cols2groups.items()]
    df_new.columns = pd.MultiIndex.from_tuples(tuples, names=["col_group", "col_name"])
    return df_new


def save_df(df, path, sheet_name="Sheet1", dates_to_str=True, index=False, verbose=True, **kwargs):
    if dates_to_str:
        df_save = convert_dates_to_str(df)
    else:
        df_save = df

    if type(path) == pd.io.excel._xlsxwriter.XlsxWriter or path.endswith("xlsx"):
        df_save.to_excel(path, sheet_name=sheet_name, engine="openpyxl", index=index, **kwargs)
    elif path.endswith(".txt") or path.endswith(".tsv"):
        df_save.to_csv(path, index=index, sep="\t", **kwargs)
    else:
        df_save.to_csv(path, index=index, **kwargs)

    if verbose:
        print("-file saved at %s" % path, flush=True)


def read_header(path: str, prefix: str) -> list:
    if path.endswith(".gz"):
        with gzip.open(path, "rt") as file:
            header = [x for x in file.readlines() if x.startswith(prefix)]
    else:
        with open(path, "r") as file:
            header = [x for x in file.readlines() if x.startswith(prefix)]
    return header


def save_df_to_data(df, path, sheet_name="Sheet1", dates_to_str=True, index=False, verbose=True, **kwargs):
    cwd = setwd_to_data()
    save_df(df, path, sheet_name, dates_to_str, index, verbose, **kwargs)
    os.chdir(cwd)


def set_dtype_columns(df, cols, dtype):
    cols_c = list(set(df.columns).intersection(set(cols)))
    for col_c in cols_c:
        df[col_c] = df[col_c].astype(dtype)
    if dtype=="str":
        df = df.replace("nan", np.nan)
    return df


def subset_data(df: DataFrame, values: Iterable[Any]=None, col_name: str=None):
    if values is None or col_name is None:
        return df
    else:
        return df.loc[df[col_name].isin(values),:]
